File: examples/c/finite_scan.c
/*
 * finite_scan.c — PRISM-CLib example: finite scan (ch1~ch4, N samples)
 * Build: make examples
 * Run  : ./finite_scan [ip] [port] [samples]
 *        ./finite_scan 192.168.7.1 7777 64000
 */
#include "prismlib.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

int main(int argc, char *argv[])
{
    const char *ip      = (argc > 1) ? argv[1] : "192.168.7.1";
    uint16_t    port    = (argc > 2) ? (uint16_t)atoi(argv[2]) : 7777;
    uint32_t    samples = (argc > 3) ? (uint32_t)atoi(argv[3]) : 64000;
    int         num_ch  = 4;
    uint8_t     ch_mask = 0x0F;

    printf("Connecting to %s:%u ...\n", ip, port);
    prismlib_t *dev = prismlib_open(ip, port);
    if (!dev) {
        fprintf(stderr, "prismlib_open failed\n");
        return 1;
    }

    char ver[32], sn[32];
    prismlib_fw_version(dev, ver);
    prismlib_serial(dev, sn);
    printf("FW version : %s\n", ver);
    printf("Serial     : %s\n", sn);

    for (int ch = 0; ch < num_ch; ch++)
        prismlib_sens_write(dev, (uint8_t)ch, 100.0);

    int use_iepe = 0;
    printf("IEPE를 활성화하시겠습니까? [y/n]: ");
    fflush(stdout);
    char ans[8] = {0};
    if (fgets(ans, sizeof(ans), stdin) && (ans[0] == 'y' || ans[0] == 'Y')) {
        use_iepe = 1;
        for (int ch = 0; ch < num_ch; ch++)
            prismlib_iepe_write(dev, (uint8_t)ch, 1);
        printf("IEPE 안정화 대기 중 (2초)...");
        fflush(stdout);
        sleep(2);
        printf(" 완료\n");
    }

    uint32_t buf_size = samples * (uint32_t)num_ch;
    double  *buf      = (double *)malloc(buf_size * sizeof(double));
    if (!buf) {
        fprintf(stderr, "malloc failed\n");
        if (use_iepe)
            for (int ch = 0; ch < num_ch; ch++)
                prismlib_iepe_write(dev, (uint8_t)ch, 0);
        prismlib_close(dev);
        return 1;
    }

    int rc = prismlib_scan_start(dev, ch_mask, samples, OPTS_DEFAULT);
    if (rc != RESULT_SUCCESS) {
        fprintf(stderr, "scan_start failed: %s\n", prismlib_error_msg(rc));
        free(buf);
        if (use_iepe)
            for (int ch = 0; ch < num_ch; ch++)
                prismlib_iepe_write(dev, (uint8_t)ch, 0);
        prismlib_close(dev);
        return 1;
    }

    uint16_t status;
    uint32_t n_read;
    rc = prismlib_scan_read(dev, &status, (int32_t)samples, 10.0,
                            buf, buf_size, &n_read);
    if (rc != RESULT_SUCCESS)
        fprintf(stderr, "scan_read failed: %s\n", prismlib_error_msg(rc));

    uint32_t avail;
    do {
        prismlib_scan_status(dev, &status, &avail);
    } while (status & STATUS_RUNNING);

    prismlib_scan_cleanup(dev);

    if (use_iepe)
        for (int ch = 0; ch < num_ch; ch++)
            prismlib_iepe_write(dev, (uint8_t)ch, 0);

    printf("Read %u samples/ch\n", n_read);
    for (uint32_t i = 0; i < n_read && i < 5; i++) {
        printf("[%4u]", i);
        for (int ch = 0; ch < num_ch; ch++)
            printf("  ch%d=%8.4f g", ch + 1, buf[i * (uint32_t)num_ch + ch]);
        printf("\n");
    }
    if (n_read > 5) printf("  ... (%u more)\n", n_read - 5);

    char fname[64];
    time_t t = time(NULL);
    struct tm *ti = localtime(&t);
    strftime(fname, sizeof(fname), "finite_scan_%Y%m%d_%H%M%S.txt", ti);
    FILE *fp = fopen(fname, "w");
    if (fp) {
        fprintf(fp, "sample");
        for (int ch = 0; ch < num_ch; ch++)
            fprintf(fp, "\tch%d(g)", ch + 1);
        fprintf(fp, "\n");
        for (uint32_t i = 0; i < n_read; i++) {
            fprintf(fp, "%u", i);
            for (int ch = 0; ch < num_ch; ch++)
                fprintf(fp, "\t%.6f", buf[i * (uint32_t)num_ch + ch]);
            fprintf(fp, "\n");
        }
        fclose(fp);
        printf("저장 완료: %s\n", fname);
    }

    free(buf);
    prismlib_close(dev);
    return 0;
}

File: examples/python/finite_scan.py
#!/usr/bin/env python3
"""finite_scan.py — PRISM-PyLib example: finite scan (ch1~ch4, N samples)

Usage:
    python finite_scan.py [ip] [port] [samples]
    python finite_scan.py 192.168.7.1 7777 1024
"""
import sys
import time
import datetime
from prismlib import prismlib, ScanOptions, ScanStatus

NUM_CHANNELS = 4

ip      = sys.argv[1] if len(sys.argv) > 1 else "192.168.7.1"
port    = int(sys.argv[2]) if len(sys.argv) > 2 else 7777
SAMPLES = int(sys.argv[3]) if len(sys.argv) > 3 else 64000

print(f"Connecting to {ip}:{port} ...")
prism = prismlib(ip, port)
prism.open()

use_iepe = False
try:
    print(f"FW version : {prism.fw_version()}")
    print(f"Serial     : {prism.serial()}")

    for ch in range(NUM_CHANNELS):
        prism.sens_write(ch, 100.0)   # 100 mV/g accelerometer

    ans = input("IEPE를 활성화하시겠습니까? [y/n]: ").strip().lower()
    use_iepe = ans in ("y", "yes")
    if use_iepe:
        for ch in range(NUM_CHANNELS):
            prism.iepe_write(ch, True)
        print("IEPE 안정화 대기 중 (2초)...", end="", flush=True)
        time.sleep(2.0)
        print(" 완료")

    prism.scan_start(0x0F, SAMPLES, ScanOptions.DEFAULT)

    data, status = prism.scan_read(SAMPLES, timeout=10.0)

    # 유한 스캔 완료 후 수신 스레드가 RUNNING 플래그를 해제할 때까지 대기.
    # scan_read() 반환 시점과 스레드 ~RUNNING 설정 사이에 race condition이 있으므로
    # scan_cleanup() 호출 전에 반드시 확인한다.
    while prism.scan_status()[0] & ScanStatus.RUNNING:
        time.sleep(0.005)

    n_sets = len(data) // NUM_CHANNELS
    print(f"Read {n_sets} samples/ch")
    for i in range(min(n_sets, 5)):
        vals = data[i * NUM_CHANNELS:(i + 1) * NUM_CHANNELS]
        print(f"[{i:4d}] " + "  ".join(f"ch{c+1}={vals[c]:8.4f} g"
                                        for c in range(NUM_CHANNELS)))
    if n_sets > 5:
        print(f"  ... ({n_sets - 5} more)")

    fname = datetime.datetime.now().strftime("finite_scan_%Y%m%d_%H%M%S.txt")
    with open(fname, "w") as f:
        f.write("sample\t" + "\t".join(f"ch{c+1}(g)" for c in range(NUM_CHANNELS)) + "\n")
        for i in range(n_sets):
            vals = data[i * NUM_CHANNELS:(i + 1) * NUM_CHANNELS]
            f.write(f"{i}\t" + "\t".join(f"{v:.6f}" for v in vals) + "\n")
    print(f"저장 완료: {fname}")

finally:
    prism.scan_cleanup()
    if use_iepe:
        for ch in range(NUM_CHANNELS):
            prism.iepe_write(ch, False)
    prism.close()
